In [1]:
import pandas as pd
import numpy as np

In [2]:
data=pd.read_csv('Data/gradedata.csv')

In [3]:
data['grade'].mean()

82.55605000000001

In [4]:
data['grade'].std()

9.74759258371467

In [5]:
data.describe()

age     exercise        hours        grade
count  2000.000000  2000.000000  2000.000000  2000.000000
mean     16.578500     3.000500    10.988500    82.556050
std       1.696254     1.423205     4.063942     9.747593
min      14.000000     0.000000     0.000000    32.000000
25%      15.000000     2.000000     8.000000    75.575000
50%      17.000000     3.000000    11.000000    82.700000
75%      18.000000     4.000000    14.000000    89.700000
max      19.000000     5.000000    20.000000   100.000000

In [6]:
data['grade'].mode()

0    100.0
dtype: float64

In [7]:
data['grade'].var()

95.01556117808924

In [8]:
data.var()

age          2.877276
exercise     2.025513
hours       16.515626
grade       95.015561
dtype: float64

In [9]:
# zbroj muških studenata

In [10]:
data.loc[data['gender']=='male'].count()

fname       1000
lname       1000
gender      1000
age         1000
exercise    1000
hours       1000
grade       1000
address     1000
dtype: int64

In [11]:
data.loc[data['grade'] > 90]

fname      lname  gender  age  exercise  hours  grade  \
7       Armando    Hoffman    male   17         5     18   95.1   
17        Libby     Guzman  female   19         1     19  100.0   
21        Maggy  Whitfield  female   15         1     15   90.5   
23        Lance   Benjamin    male   14         5     18   90.3   
26      Solomon  Mcpherson    male   15         5     18   94.5   
29         Emma     Mccall  female   16         2     13   91.1   
30       Isaiah     Newman    male   16         2     12   91.9   
32      Georgia      Munoz  female   14         5     20   99.0   
33        Doris   Melendez  female   17         5     12   93.2   
36         Ivor     Arnold    male   19         4     20  100.0   
37        Devin     Jacobs    male   14         3     14   98.1   
43        Hilel   Hamilton    male   15         2     17   95.2   
45         Rhea     Melton  female   18         5     12   95.0   
69        Daryl     Benson  female   19         1     18   93.8   
71    Christine     Briggs  female   14         5     18   91.9   
77       Quemby    Justice  female   14         2     17  100.0   
85       Porter       Ware    male   14         1     16   90.4   
86         Joan     Sexton  female   18         5     18   96.6   
90       Sydnee   Campbell  female   18         2     15   99.4   
92       Alyssa      Craig  female   19         5     11   93.0   
97      Cassidy     Hughes  female   15         3     17   92.0   
100     Gabriel     Morton    male   16         2     15   91.3   
101        Sage  Cleveland  female   19         2     20  100.0   
105      Chaney    Hoffman    male   14         3     13   93.6   
106       Fritz       Park    male   14         2     18   93.2   
108        Lara      Haley  female   17         5     12   95.6   
109      Sophia     Gordon  female   19         4     17  100.0   
110        Kaye     Howard  female   19         5     13   94.9   
112     Xanthus     Barlow    male   19         5     16   95.5   
113    Hamilton    Terrell    male   15         5     16   96.6   
...         ...        ...     ...  ...       ...    ...    ...   
1893     Juliet       Good  female   14         5     15  100.0   
1897     Hakeem       Roth    male   17         5     10   90.6   
1899       Nora     Gaines  female   19         3     15   98.9   
1903    Phyllis      Moore  female   18         2     14   92.4   
1904      Vance       Byrd    male   18         4     17   91.5   
1907    Zenaida       Page  female   17         4     18   95.6   
1911       Neil    Collins    male   17         4     12   94.9   
1919       Mari    Collins  female   18         3     15   97.5   
1922     Phelan       Frye    male   16         5     19  100.0   
1926    Lillith    Bradley  female   16         2     19   92.4   
1933      Adena   Robinson  female   14         3     19  100.0   
1936       Orla     Waller  female   18         4     17   91.3   
1938   Demetria   Espinoza  female   16         5     17   92.5   
1945    Brianna   Holloway  female   18         4     18  100.0   
1949    Armando    Mccarty    male   15         3     13   93.8   
1950     Kaseem      Finch    male   18         1     16   97.1   
1955     Carlos     Ingram    male   16         5     17   98.7   
1957     Aileen     Sutton  female   18         5     16   91.4   
1959    Chelsea   Holloway  female   19         4     12   91.4   
1964    Zenaida    Maynard  female   16         2     19   91.3   
1967     Aurora      Gould  female   15         3     18   90.3   
1970    Armando    Mcclure    male   19         2     14   93.2   
1971      Haley    Mcgowan  female   17         3     16   90.4   
1979        Kai    Woodard  female   17         5     17  100.0   
1980     Maxine    Raymond  female   17         4     19   91.9   
1981       Leah   Lawrence  female   15         5     15  100.0   
1983    Anthony     Palmer    male   14         1     14   93.1   
1986     Indigo      Mccoy  female   19         2     14 

compare average points of male and female students

In [12]:
data.loc[data['gender']=='female']['grade'].mean()

82.71730000000002

In [13]:
data.loc[data['gender']=='male']['grade'].mean()

82.3948

Sort data by age desc.

In [14]:
data=data.sort_values(by='age', ascending=0)

In [15]:
data.head(3)

fname   lname  gender  age  exercise  hours  grade  \
1000    Hanna  Mooney  female   19         2     10   88.8   
902    Steven  Graham    male   19         4     17   95.1   
818   Colette   Rojas  female   19         4      6   73.3   

                                         address  
1000  8293 SW. Cedar Swamp Lane, Union, NJ 07083  
902                90 Grove St., Camas, WA 98607  
818      17 Pin Oak Street, Villa Park, IL 60181

In [16]:
data.tail(3)

fname    lname  gender  age  exercise  hours  grade  \
588  Jocelyn   Ashley  female   14         3      9   73.4   
89    Lucius  Herrera    male   14         2     19   88.0   
294  Illiana   Branch  female   14         5      4   69.8   

                                        address  
588      830 Amerige Drive, Davenport, IA 52804  
89           67 Old Clinton Dr., Lynn, MA 01902  
294  9730 Van Dyke Court, Port Orange, FL 32127

Sorting by hours of study and exercise, ascending

In [17]:
data=data.sort_values(by=['hours','exercise'], ascending=[True, True])

In [18]:
data.head(5)

fname    lname gender  age  exercise  hours  grade  \
356      Luke  Santana   male   17         1      0   75.0   
1611  Garrett  Everett   male   17         2      0   78.1   
1213   Arthur   Nieves   male   14         3      0   76.8   
1507   Ferris    Avila   male   18         5      1   74.2   
80       Hoyt    Lyons   male   16         5      1   71.3   

                                               address  
356                2 Sunbeam Dr., Zionsville, IN 46077  
1611       7963 South Thorne Drive, Sterling, VA 20164  
1213          377 Beech Street, Chesterfield, VA 23832  
1507               85 Brown Drive, Pensacola, FL 32503  
80    720 Linden Street, Springfield Gardens, NY 11413

# Correlation

In [19]:
data.corr()

age  exercise     hours     grade
age       1.000000 -0.003643 -0.017467 -0.007580
exercise -0.003643  1.000000  0.021105  0.161286
hours    -0.017467  0.021105  1.000000  0.801955
grade    -0.007580  0.161286  0.801955  1.000000

hours of study and grade are the high correlated

First Regression

In [20]:
import statsmodels.formula.api as sm

In [21]:
result=sm.ols(formula='grade~age + exercise + hours',data=data).fit()
result.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  grade   R-squared:                       0.664
Model:                            OLS   Adj. R-squared:                  0.664
Method:                 Least Squares   F-statistic:                     1315.
Date:                Wed, 01 May 2019   Prob (F-statistic):               0.00
Time:                        10:17:49   Log-Likelihood:                -6300.7
No. Observations:                2000   AIC:                         1.261e+04
Df Residuals:                    1996   BIC:                         1.263e+04
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept     57.8704      1.321     43.804      0.000      55.279      60.461
age            0.0397      0.075      0.532      0.595      -0.107       0.186
exercise       0.9893      0.089     11.131      0.000       0.815       1.164
hours          1.9165      0.031     61.564      0.000       1.855       1.978
==============================================================================
Omnibus:                      321.187   Durbin-Watson:                   1.872
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             2196.187
Skew:                          -0.567   Prob(JB):                         0.00
Kurtosis:                       8.007   Cond. No.                         213.
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

Comment:

* R-squared - percentage of the variation in the data that can be accounted for by the regression (664 is good but not great)
* p-value (P > |t|) - represents the probability that the independet variable has no effect on the dependent variable - keep p- values under 5 %

In [22]:
#rerun regression without age

In [23]:
result2=sm.ols(formula='grade ~ exercise + hours', data=data).fit()
result2.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  grade   R-squared:                       0.664
Model:                            OLS   Adj. R-squared:                  0.664
Method:                 Least Squares   F-statistic:                     1973.
Date:                Wed, 01 May 2019   Prob (F-statistic):               0.00
Time:                        10:17:49   Log-Likelihood:                -6300.8
No. Observations:                2000   AIC:                         1.261e+04
Df Residuals:                    1997   BIC:                         1.262e+04
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept     58.5316      0.447    130.828      0.000      57.654      59.409
exercise       0.9892      0.089     11.131      0.000       0.815       1.163
hours          1.9162      0.031     61.575      0.000       1.855       1.977
==============================================================================
Omnibus:                      318.721   Durbin-Watson:                   1.872
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             2158.000
Skew:                          -0.564   Prob(JB):                         0.00
Kurtosis:                       7.962   Cond. No.                         43.2
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

* We haven't changed R squared, but we changed p-values. So, we can now look at our coefficients.
* grade=1.916 hours + .989 hours of exercise + 58.5316

In [24]:
from sklearn.preprocessing import LabelEncoder

In [25]:
le_gender=LabelEncoder()

In [27]:
data['gender']=le_gender.fit_transform(data['gender'])

In [28]:
data['gender']

356     1
1611    1
1213    1
1507    1
80      1
1278    1
700     1
556     0
1317    0
1110    0
1359    1
903     1
1640    1
836     1
393     1
1931    1
1449    0
410     0
1916    1
1758    1
951     1
1210    1
1586    1
1494    0
1712    0
838     0
1207    0
864     1
1910    0
1512    1
       ..
1029    0
321     0
621     0
752     1
1184    0
1922    1
899     1
1269    1
1070    1
1144    0
1481    0
101     0
1719    0
1733    0
252     0
226     1
1665    1
1889    1
204     1
36      1
1021    0
1650    0
164     1
954     0
508     0
1356    1
616     1
690     1
825     1
32      0
Name: gender, Length: 2000, dtype: int64

Can be R squared be improved?

In [29]:
result3=sm.ols(formula='grade ~ gender + exercise + hours', data=data).fit()
result3.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  grade   R-squared:                       0.665
Model:                            OLS   Adj. R-squared:                  0.664
Method:                 Least Squares   F-statistic:                     1318.
Date:                Wed, 01 May 2019   Prob (F-statistic):               0.00
Time:                        10:19:50   Log-Likelihood:                -6299.3
No. Observations:                2000   AIC:                         1.261e+04
Df Residuals:                    1996   BIC:                         1.263e+04
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept     58.7621      0.466    126.172      0.000      57.849      59.676
gender        -0.4476      0.253     -1.770      0.077      -0.943       0.048
exercise       0.9840      0.089     11.073      0.000       0.810       1.158
hours          1.9170      0.031     61.628      0.000       1.856       1.978
==============================================================================
Omnibus:                      322.999   Durbin-Watson:                   1.872
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             2244.206
Skew:                          -0.566   Prob(JB):                         0.00
Kurtosis:                       8.064   Cond. No.                         45.6
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

# Pivot Tables

In [30]:
# average by gender

In [34]:
pd.pivot_table(data, index=['gender'])

age  exercise    grade   hours
gender                                   
0       16.568     3.047  82.7173  10.932
1       16.589     2.954  82.3948  11.045

In [35]:
#average grade by gender

In [36]:
pd.pivot_table(data, values=['grade'], index=['gender'])

grade
gender         
0       82.7173
1       82.3948

In [37]:
#min grade by gender

In [38]:
pd.pivot_table(data, values=['grade'], index=['gender'], aggfunc='min')

grade
gender       
0        32.0
1        43.0

In [39]:
#max grade by gender and age

In [40]:
pd.pivot_table(data, index=['gender','age'], aggfunc='max', values=['hours'])

hours
gender age       
0      14      20
       15      20
       16      19
       17      20
       18      20
       19      20
1      14      19
       15      20
       16      20
       17      20
       18      20
       19      20

In [41]:
#average grade and hours by gender

In [43]:
pd.pivot_table(data, values=['grade','hours'], index='gender', aggfunc='mean')

grade   hours
gender                 
0       82.7173  10.932
1       82.3948  11.045

In [44]:
# find average grade and hours of study by gender for students who are 17 y

In [45]:
data_s=data[data['age']==17]

In [47]:
data_s.head()

fname     lname  gender  age  exercise  hours  grade  \
356      Luke   Santana       1   17         1      0   75.0   
1611  Garrett   Everett       1   17         2      0   78.1   
1317   Kessie  Mcdonald       0   17         1      2   62.2   
1110   Xerxes    Mooney       0   17         1      2   70.4   
1586    Burke   Bennett       1   17         1      3   65.8   

                                          address  
356           2 Sunbeam Dr., Zionsville, IN 46077  
1611  7963 South Thorne Drive, Sterling, VA 20164  
1317             311 Pierce St., Eugene, OR 97402  
1110            974 Theatre Dr., Arvada, CO 80003  
1586  707 Schoolhouse Court, Ridgefield, CT 06877

In [48]:
pd.pivot_table(data_s, values=['grade','hours'], index=['gender'], aggfunc='mean')

grade      hours
gender                      
0       83.599435  10.943503
1       82.949721  11.268156

In [49]:
#including totals in average grade and hours by gender

In [50]:
pd.pivot_table(data_s, values=['grade','hours'], index=['gender'], aggfunc='mean',margins=True)

grade      hours
gender                      
0       83.599435  10.943503
1       82.949721  11.268156
All     83.272753  11.106742

In [51]:
#create pivot table showing average grade by gender of people who had more than 2 h of exercise

In [52]:
data_e=data[data['exercise'] > 2]

In [53]:
pd.pivot_table(data_e, values=['grade'],index=['gender'], aggfunc='mean')

grade
gender           
0       83.730343
1       83.622542